models/domains_tlds.php

```diff
@@ -414,41 +429,201 @@ public function edit($tld, array $vars)
+            return false;
+        }
+
+        // Check if there are any services using this module and package
+        $services = $this->Record->select(['services.*', 'packages.id' => 'package_id'])
+            ->from('services')
+            ->innerJoin('package_pricing', 'package_pricing.id', '=', 'services.pricing_id', false)
+            ->innerJoin('packages', 'packages.id', '=', 'package_pricing.package_id', false)
+            ->where('packages.id', '=', $package_id)
+            ->numResults();
+
+        return ($services > 0);
+    }
+
+    /**
+     * Migrate the TLD to a new module
+     *
+     * @param string $tld The TLD to migrate
+     * @param int $new_module_id The ID of the new module for the TLD
+     */
+    private function migrateModule($tld, $new_module_id)
+    {
+        Loader::loadModels($this, ['Packages']);
+
+        $tld = $this->get($tld);
+
+        // Get old package
+        $old_package_id = isset($tld->package_id) ? $tld->package_id : null;
+        $old_package = $this->Packages->get($old_package_id);
+
+        // Check if there is any existing package using the new module
+        $package = $this->getTldPackageByModuleId($tld->tld, $new_module_id);
+
+        if (!empty($package)) {
+            $package_id = $package->id;
+
+            // Set the current status to the restored package
+            $this->Packages->edit($package_id, ['status' => $old_package->status]);
+        } else {
+            // Create a new package with the same tld and using the new module
+            $params = [
+                'module_id' => $new_module_id,
+                'tld' => $tld->tld,
+                'status' => $old_package->status
+            ];
+            $package_id = $this->createPackage($params);
+
+            // Set the old meta data and pricing to the new package
+            $params = [
+                'pricing' => (isset($old_package->pricing) ? $old_package->pricing : []),
+                'meta' => (isset($old_package->meta) ? $old_package->meta : [])
+            ];
+            $params = json_decode(json_encode($params), true);
+            $this->Packages->edit($package_id, $params);
+
+            if (($errors = $this->Packages->errors())) {
+                $this->Input->setErrors($errors);
+
+                return;
+            }
+
+            // Assign the new package to the TLD
+            $this->Record->insert('domains_packages', ['tld_id' => $tld->id, 'package_id' => $package_id]);
+        }
+
+        // Set the status of the old package as inactive
+        $this->Record->duplicate('domains_packages.package_id', '=', $old_package_id)
+            ->insert('domains_packages', ['tld_id' => $tld->id, 'package_id' => $old_package_id]);
+        $this->Packages->edit($old_package_id, ['status' => 'inactive']);
+
+        // Update TLD
+        $this->edit($tld->tld, ['package_id' => $package_id]);
```

Comment:
This is already part of a call to ::edit().  Perhaps we can just return the $package_id and in ::edit() do $vars['package_id'] = $this->migrateModule(...);